test/core/promise/promise_fuzzer.cc

```diff
@@ -12,10 +12,306 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include "src/core/lib/promise/activity.h"
+#include "src/core/lib/promise/join.h"
+#include "src/core/lib/promise/map.h"
+#include "src/core/lib/promise/promise.h"
+#include "src/core/lib/promise/race.h"
+#include "src/core/lib/promise/seq.h"
 #include "src/libfuzzer/libfuzzer_macro.h"
 #include "test/core/promise/promise_fuzzer.pb.h"
 
 bool squelch = true;
 bool leak_check = true;
 
-DEFINE_PROTO_FUZZER(const promise_fuzzer::Msg&) {}
+namespace grpc_core {
+// Return type for infallible promises.
+// We choose this so that it's easy to construct, and will trigger asan failures
+// if misused, and is copyable.
+using IntHdl = std::shared_ptr<int>;
+
+template <typename T>
+using PromiseFactory = std::function<Promise<T>(T)>;
+
+namespace {
+class Fuzzer {
+ public:
+  void Run(const promise_fuzzer::Msg& msg) {
+    // If there's no promise we can't construct and activity and... we're done.
+    if (!msg.has_promise()) {
+      return;
+    }
+    // Construct activity.
+    activity_ = MakeActivity(
+        [msg, this] {
+          return Seq(MakePromise(msg.promise()),
+                     [] { return absl::OkStatus(); });
+        },
+        Scheduler{this},
+        [this](absl::Status status) {
+          // Must only be called once
+          GPR_ASSERT(!done_);
+          // If we became certain of the eventual status, verify it.
+          if (expected_status_.has_value()) {
+            GPR_ASSERT(status == *expected_status_);
+          }
+          // Mark ourselves done.
+          done_ = true;
+        });
+    for (int i = 0; !done_ && activity_ != nullptr && i < msg.actions_size();
+         i++) {
```

Comment:
super-nit: prefer prefix increment per the style guide, unless you need postfix increment behavior.